### PR TITLE
Store step population and best ever solutions in CMA-ES

### PR DESCRIPTION
**Pull request by egomeh**
* In addition to the best solution in a `step`, make the CMA-ES store all population values from a `step`.
* Track best ever solution on the CMA-ES.
* Some changes in style in comments and code for better consistency.
* Add interface for evolution strategy algorithms.

### CODE REVIEW

File: include/shark/Algorithms/DirectSearch/CMA.h [hunk 1]
```diff
@@ -204,6 +204,20 @@ class CMA : public AbstractSingleObjectiveOptimizer<RealVector >
 		return m_numEvaluations;
 	}
 
+	///\breif Return the population form the previous step
+	std::vector<SolutionType> population() const {
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/CMA.h]: return type should be const&

File: include/shark/Algorithms/DirectSearch/CMA.h [hunk 2]
```diff
@@ -204,6 +204,20 @@ class CMA : public AbstractSingleObjectiveOptimizer<RealVector >
...
+		return m_population;
+	}
+
+	///\brief Returns the best solution of the previous step
+	SolutionType best() const {
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/CMA.h]: this is already implemented in the optimizer base class

File: include/shark/Algorithms/DirectSearch/CMA.h [hunk 3]
```diff
@@ -204,6 +204,20 @@ class CMA : public AbstractSingleObjectiveOptimizer<RealVector >
...
+		return m_best;
+	}
+
+	///\brief Returns the best ever discovered solution
+	SolutionType bestEver() const {
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/CMA.h]: If we want that, this should also go in the base class. But I am not sure that we want that. Often, what is considered the "all-time-best" solution would depend on a different metric, e.g. classification error of the model.

File: include/shark/Algorithms/AbstractEvolutionStrategy.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+/// \brief An optimizer that applies an evolution stragtegy
+/// 
+/// This interface serves the purpose of allowing an evolution strategy based
+/// optimizer to supply a population set to an external evaluation mechanism
+/// as opposed to only expose the `step` function. This puts more control
+/// in the user of the optimizer as evaluation is shifted to be fullly controlled
+/// by the user of the optimizer instead of having the optimizer handle evaluation
+/// internally.
+///
+/// Also when init() is called as offered by the AbstractOptimizer interface, the function
```
Comment (Ulfgard) [inline include/shark/Algorithms/AbstractEvolutionStrategy.h]: this is a global requirement. optimizer.init(f); requires that f can propose a starting point. See
https://github.com/Shark-ML/Shark/blob/d63962161d9e7229c804373f812015f84a121139/include/shark/Algorithms/AbstractOptimizer.h#L118

File: include/shark/Algorithms/DirectSearch/CMA.h (hunk 1 above)
Comment (egomeh) [reply include/shark/Algorithms/DirectSearch/CMA.h]: Ah yes, I missed that.

File: include/shark/Algorithms/DirectSearch/CMA.h (hunk 2 above)
Comment (egomeh) [reply include/shark/Algorithms/DirectSearch/CMA.h]: Alright, I'll remove it in a PR update.

File: include/shark/Algorithms/DirectSearch/CMA.h (hunk 3 above)
Comment (egomeh) [reply include/shark/Algorithms/DirectSearch/CMA.h]: I see your point. I will remove these in a PR update. If the `best` ever is to be tracked, one might as well do this manually as the `best` is exposed per iteration.

**Comment by egomeh** [comment]
I have updated the CMA-ES to inherit the `AbstractEvolutionStrategy` and make the sampling function and distribution update functions public. This would make it fairly easy to apply a custom selection mechanism.

**Comment by Ulfgard** [comment]
Hi,

I will for now not accept your pull-request. we wait until your project is done and until then you can work in your own branch. i will pull in the end. I have to see everything in a broader perspective than only EA, so I might make my final design decisions slightly differently.


________________________________

**Pull request closed without merging**